### PR TITLE
gnatboot: 4.1 -> 11.2.0-4

### PR DESCRIPTION
**Pull request by ethindp**
###### Description of changes

This PR does a couple things:

* All gnat builds now use the gnatboot package instead of a previous gnat version.
* The "gnat" package has been changed to reference "gnat12".
* The gnatboot package now uses prebuilt binaries provided by the [Alire project](https://github.com/alire-project/gnat-fsf-builds).

Notes:

1. GNAT 6 and 9 do not build successfully. Build failures will continue to occur; this should, however, be resolvable if someone really needs these old versions.
2. As of now, the gnatboot package uses x86-64-specific binaries. No other architectures are supported. Cross-compilation is possible, but will need to be bootstrapped from an x86-64 system.
3. The aforementioned repository does contain cross-compilers for other architectures; however, these are x86-64-specific, so it seemed redundant to include them, as we can simply reproduce them in nixpkgs automatically.

Suggestions:

Due to the continuing build failures with GCC versions before 10.x.x, I suggest that we simply remove GCC versions older than 10.x.x. In particular, removal of GNAT 6 and 9 is strongly encouraged, as it is very unlikely that anyone
will actually need to use these very old versions of Gnat.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 177579` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by ethindp** [comment]
@siraben Please clarify... I found this to be confusing.

**Comment by siraben** [comment]
@ethindp it appears there are no rebuilds on `aarch64-darwin`.

File: pkgs/development/compilers/gnatboot/default.nix [hunk 1]
```diff
@@ -1,55 +1,106 @@
...
-
-  dontStrip = 1;
+{ stdenv, lib, autoPatchelfHook, fetchzip, lzma, ncurses5, readline, gmp, mpfr
+, expat, libipt, zlib, dejagnu, sourceHighlight, python3, elfutils, guile, glibc
+}:
 
-  installPhase = ''
-    mkdir -p $out
-    cp -R * $out
+stdenv.mkDerivation rec {
```
Comment (Mindavi) [inline pkgs/development/compilers/gnatboot/default.nix:5]: Maybe this should have an assert for the platform?

Comment (Mindavi) [reply pkgs/development/compilers/gnatboot/default.nix:5]: Duh no, the platforms should be updated instead :)

**Comment by ethindp** [comment]
@Mindavi Looking at the derivation it already does in the metadata...

**Comment by ethindp** [comment]
At least for gnatboot...

File: pkgs/development/compilers/gnatboot/default.nix [hunk 2]
```diff
@@ -1,55 +1,106 @@
...
   meta = with lib; {
-    homepage = "https://gentoo.org";
-    license = licenses.gpl3Plus;
-    maintainers = [ maintainers.lucus16 ];
-
+    description = "GNAT, the GNU Ada Translator";
+    homepage = "https://www.gnu.org/software/gnat";
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ ethindp ];
     platforms = platforms.linux;
```
Comment (Mindavi) [inline pkgs/development/compilers/gnatboot/default.nix:104]: Heh, definitely not linux-x86_64 right?

**Comment by Mindavi** [comment]
Maybe I'm mislooking, but definitely looks like it isn't constrained

File: pkgs/development/compilers/gnatboot/default.nix [hunk 3]
```diff
@@ -1,55 +1,106 @@
...
+    sourceHighlight
+    python3
+    elfutils
+    guile
+    glibc
+  ];
+
+  installPhase = ''
+    mkdir -p $out
+    cp -var * $out/
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gnatboot/default.nix:89]:
```suggestion
    cp -ar * $out/
```

File: pkgs/development/compilers/gnatboot/default.nix [hunk 4]
```diff
@@ -1,55 +1,106 @@
...
+    expat
+    libipt
+    zlib
+    dejagnu
+    sourceHighlight
+    python3
+    elfutils
+    guile
+    glibc
+  ];
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gnatboot/default.nix:49]: Please properly sort the dependencies. Having all of them in all inputs is most likely not the minimal dependencies.

File: pkgs/development/compilers/gnatboot/default.nix [hunk 5]
```diff
@@ -1,55 +1,106 @@
...
-  installPhase = ''
-    mkdir -p $out
-    cp -R * $out
+stdenv.mkDerivation rec {
+  pname = "gnatboot";
+  version = "11.2.0-4";
+  src = fetchzip {
+    url =
+      "https://github.com/alire-project/GNAT-FSF-builds/releases/download/gnat-"
+      + version + "/gnat-x86_64-linux-" + version + ".tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/gnatboot/default.nix:11]:
```suggestion
  version = "11.2.0-4";

  src = fetchzip {
    url = "https://github.com/alire-project/GNAT-FSF-builds/releases/download/gnat-${version}/gnat-x86_64-linux-${version}.tar.gz";
```

File: pkgs/development/compilers/gnatboot/default.nix (hunk 4 above)
Comment (ethindp) [reply pkgs/development/compilers/gnatboot/default.nix:49]: I'll sort them, but I'm pretty sure all of them are required.

**Comment by ethindp** [comment]
I can't seem to squash the three commits into fe84e15b3a6f1e87b8b0fd04c84f33e3dae2a7ad for some reason. Apologies.

File: pkgs/development/compilers/gnatboot/default.nix (hunk 4 above)
Comment (ethindp) [reply pkgs/development/compilers/gnatboot/default.nix]: Got it. Thanks.

File: pkgs/development/compilers/gnatboot/default.nix [hunk 6]
```diff
@@ -1,55 +1,51 @@
...
-    }
-  else
-    throw "Platform not supported";
+  src = fetchzip {
+    url = "https://github.com/alire-project/GNAT-FSF-builds/releases/download/gnat-${version}/gnat-x86_64-linux-${version}.tar.gz";
+    hash = "sha256-8fMBJp6igH+Md5jE4LMubDmC4GLt4A+bZG/Xcz2LAJQ=";
+  };
 
-  dontStrip = 1;
+  propagatedNativeBuildInputs = [
```
Comment (Mindavi) [inline pkgs/development/compilers/gnatboot/default.nix:14]: Why are these all propagated? For the previous gnatboot it didn't seem necessary to propagate them, right?

File: pkgs/development/compilers/gnatboot/default.nix [hunk 7]
```diff
@@ -1,55 +1,51 @@
...
-    set -e
-
-    mv $out/bin/gnatgcc_2wrap $out/bin/gnatgcc
-    ln -s $out/bin/gnatgcc $out/bin/gcc
+    cp -ar * $out/
   '';
 
   passthru = {
-    langC = true; # TRICK for gcc-wrapper to wrap it
+    langC = true;
```
Comment (Mindavi) [inline pkgs/development/compilers/gnatboot/default.nix:38]: Why remove the comment? It's still a hack

Comment (ethindp) [reply pkgs/development/compilers/gnatboot/default.nix]: I've restored it -- point taken.

File: pkgs/development/compilers/gnatboot/default.nix (hunk 6 above)
Comment (Mindavi) [reply pkgs/development/compilers/gnatboot/default.nix:14]: autoPatchElfHook only operates on this derivation right? I'd expect those deps in nativeBuildInputs should be enough if they're all patched into the RPATH

Comment (ethindp) [reply pkgs/development/compilers/gnatboot/default.nix]: I tried that in the initial testing of it before I sent this PR in and it didn't work properly for some reason. I altered `default.nix` as suggested and it seems to work this time... Not sure what changed.

**Review by Mindavi — CHANGES_REQUESTED** [review]
I got around to testing this, but in general it seems like some work is needed:

<details><summary>Logs of building reverse dependencies</summary>

```
error: builder for '/nix/store/bk04ngjy9s84fgib3x5aqsdwzfhps0jq-gnat-9.3.0.drv' failed with exit code 77;
       last 10 log lines:
       > checking for libitm support... yes
       > checking for libsanitizer support... yes
       > checking for libvtv support... yes
       > checking for libhsail-rt support... yes
       > checking for libphobos support... yes
       > checking for x86_64-unknown-linux-gnu-gcc... gcc
       > checking whether the C compiler works... no
       > configure: error: in `/build/build':
       > configure: error: C compiler cannot create executables
       > See `config.log' for more details
       For full logs, run 'nix log /nix/store/bk04ngjy9s84fgib3x5aqsdwzfhps0jq-gnat-9.3.0.drv'.
error: builder for '/nix/store/2limig9gj7akjb645326vfr9yrn46sm8-gnat-6.5.0.drv' failed with exit code 77;
       last 10 log lines:
       > checking for libcilkrts support... yes
       > checking for libitm support... yes
       > checking for libsanitizer support... yes
       > checking for libvtv support... yes
       > checking for libmpx support... yes
       > checking for x86_64-unknown-linux-gnu-gcc... gcc
       > checking for C compiler default output file name...
       > configure: error: in `/build/build':
       > configure: error: C compiler cannot create executables
       > See `config.log' for more details.
       For full logs, run 'nix log /nix/store/2limig9gj7akjb645326vfr9yrn46sm8-gnat-6.5.0.drv'.
error: builder for '/nix/store/35qdwxz9psc3nmqrfjydsnlqb0nn4sy2-gnat-11.3.0.drv' failed with exit code 77;
       last 10 log lines:
       > checking for libitm support... yes
       > checking for libsanitizer support... yes
       > checking for libvtv support... yes
       > checking for libhsail-rt support... yes
       > checking for libphobos support... yes
       > checking for x86_64-unknown-linux-gnu-gcc... gcc
       > checking whether the C compiler works... no
       > configure: error: in `/build/build':
       > configure: error: C compiler cannot create executables
       > See `config.log' for more details
       For full logs, run 'nix log /nix/store/35qdwxz9psc3nmqrfjydsnlqb0nn4sy2-gnat-11.3.0.drv'.
error: 1 dependencies of derivation '/nix/store/phcpfk16i1l0jj0p6vsdmasp7v4g20j2-gnat-wrapper-11.3.0.drv' failed to build
error: builder for '/nix/store/vpar5zx7psm6hjwcfz0s5mbz3izblwhv-gnat-12.1.0.drv' failed with exit code 77;
       last 10 log lines:
       > checking for libatomic support... yes
       > checking for libitm support... yes
       > checking for libsanitizer support... yes
       > checking for libvtv support... yes
       > checking for libphobos support... yes
       > checking for x86_64-unknown-linux-gnu-gcc... gcc
       > checking whether the C compiler works... no
       > configure: error: in `/build/build':
       > configure: error: C compiler cannot create executables
       > See `config.log' for more details
       For full logs, run 'nix log /nix/store/vpar5zx7psm6hjwcfz0s5mbz3izblwhv-gnat-12.1.0.drv'.
```

</details>

Am I doing something wrong here? I did this: `nix build .#gnatboot .#gnat6 .#gnat9 .#gprbuild .#xmlada .#coreboot-toolchain.arm .#ghdl-mcode --keep-going --max-jobs 2` where I hand-picked some random things from the ofborg eval list that seemed relevant.

**Comment by ethindp** [comment]
@Mindavi This does not build with gnat6 or 9. Gnat 6 and 9 are still broken; only gnat 10/11/12 are supported (though gnat10 may not build either). This is not surprising -- its unlikely that you could build, say, GCC 3.x with GCC 12.x.

**Comment by Mindavi** [comment]
> @Mindavi This does not build with gnat6 or 9. Gnat 6 and 9 are still broken; only gnat 10/11/12 are supported (though gnat10 may not build either). This is not surprising -- its unlikely that you could build, say, GCC 3.x with GCC 12.x.

Sure, but 11 and 12 were also failing to build (I edited those in after re-reading the initial PR message, since I just picked the first few fails before).

File: pkgs/development/compilers/gnatboot/default.nix [hunk 8]
```diff
@@ -1,41 +1,37 @@
...
-  else if stdenv.hostPlatform.system == "x86_64-linux" then
-    fetchurl {
-      url = "mirror://gentoo/distfiles/gnatboot-4.1-amd64.tar.bz2";
-      sha256 = "1li4d52lmbnfs6llcshlbqyik2q2q4bvpir0f7n38nagp0h6j0d4";
-    }
-  else
-    throw "Platform not supported";
+  src = fetchzip {
+    url = "https://github.com/alire-project/GNAT-FSF-builds/releases/download/gnat-${version}/gnat-x86_64-linux-${version}.tar.gz";
+    hash = "sha256-8fMBJp6igH+Md5jE4LMubDmC4GLt4A+bZG/Xcz2LAJQ=";
```
[the remark this answers is not in the document]
Comment (ethindp) [reply pkgs/development/compilers/gnatboot/default.nix:11]: This is quite odd; it doesn't happen on my machine at all.

**Comment by ethindp** [comment]
@Mindavi I'm doing that now, and it seems to be working just fine. No hash mismatches on my end... I'm very confused now.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 177579` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by Mindavi** [comment]
Result of `nixpkgs-review pr 177579` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>coreboot-toolchain.nds32le</li>
    <li>ghdl</li>
    <li>ghdl-llvm</li>
    <li>gnat6</li>
    <li>gnat9</li>
    <li>yosys-ghdl</li>
  </ul>
</details>
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>coreboot-toolchain.aarch64</li>
    <li>coreboot-toolchain.arm</li>
    <li>coreboot-toolchain.i386</li>
    <li>coreboot-toolchain.ppc64</li>
    <li>coreboot-toolchain.riscv</li>
    <li>coreboot-toolchain.x64</li>
    <li>gnat</li>
    <li>gnat10</li>
    <li>gnat11</li>
    <li>gnatboot</li>
    <li>gnatcoll-core</li>
    <li>gnatcoll-db2ada</li>
    <li>gnatcoll-gmp</li>
    <li>gnatcoll-iconv</li>
    <li>gnatcoll-lzma</li>
    <li>gnatcoll-omp</li>
    <li>gnatcoll-postgres</li>
    <li>gnatcoll-python3</li>
    <li>gnatcoll-readline</li>
    <li>gnatcoll-sql</li>
    <li>gnatcoll-sqlite</li>
    <li>gnatcoll-syslog</li>
    <li>gnatcoll-xref</li>
    <li>gnatcoll-zlib</li>
    <li>gnatinspect</li>
    <li>gprbuild</li>
    <li>gprbuild-boot</li>
    <li>xmlada</li>
  </ul>
</details>

**Comment by Mindavi** [comment]
Would be nice if you could squash the commits together to be more logical units, that'll reduce issues for anyone bisecting in the future.

**Review by Mindavi — APPROVED** [review]

**Comment by ethindp** [comment]
@Mindavi I'm unsure why you've got those package failures (gnat 6.x and 9.x are explainable -- I'm not going to bother fixing those because I doubt anyone would ever need such an old version of gnat), but coreboot-toolchain.nds32le, ghdl, ghdl-llvm, and yosys-ghdl are confusing and I'll investigate them.
I can't seem to squash the commits; for some reason Git won't allow me to (and I'm not an expert with git rebasing so I might need some help with the command to get it to let me do that). Normally `git rebase -i` lets me do it but it just gives `noop` so... :confused:

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 177579` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>ghdl (ghdl-mcode)</li>
    <li>ghdl-llvm</li>
    <li>gnat6</li>
    <li>gnat9</li>
    <li>yosys-ghdl</li>
  </ul>
</details>
<details>
  <summary>29 packages built:</summary>
  <ul>
    <li>coreboot-toolchain.aarch64</li>
    <li>coreboot-toolchain.arm</li>
    <li>coreboot-toolchain.i386</li>
    <li>coreboot-toolchain.nds32le</li>
    <li>coreboot-toolchain.ppc64</li>
    <li>coreboot-toolchain.riscv</li>
    <li>coreboot-toolchain.x64</li>
    <li>gnat (gnat12)</li>
    <li>gnat10</li>
    <li>gnat11</li>
    <li>gnatboot</li>
    <li>gnatcoll-core</li>
    <li>gnatcoll-db2ada</li>
    <li>gnatcoll-gmp</li>
    <li>gnatcoll-iconv</li>
    <li>gnatcoll-lzma</li>
    <li>gnatcoll-omp</li>
    <li>gnatcoll-postgres</li>
    <li>gnatcoll-python3</li>
    <li>gnatcoll-readline</li>
    <li>gnatcoll-sql</li>
    <li>gnatcoll-sqlite</li>
    <li>gnatcoll-syslog</li>
    <li>gnatcoll-xref</li>
    <li>gnatcoll-zlib</li>
    <li>gnatinspect</li>
    <li>gprbuild</li>
    <li>gprbuild-boot</li>
    <li>xmlada</li>
  </ul>
</details>

**Pull request merged**